File: examples/cs/gmsub/gmsub.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/*! \file gmsub.cs
 * This file contains a simple example subscriber using the GMSEC .NET API.
 */


namespace gmsub
{
	using GMSEC.API;
	using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;
	using System.Diagnostics;
	using System.Collections.Generic;


	class gmsub : GmsecExample
	{
		private Config     config;
		private Connection conn;
        private List<SubscriptionInfo> info = new List<SubscriptionInfo>(); 

		public gmsub()
		{
		}


		private void PrintUsage()
		{
			Log.Error("\nusage: gmsub_cs connectiontype=<middleware> " +
			          "[ <parameter>=<value> ]\n" + 
			          "\n\tNote that the parameter 'connectiontype' is required. " +
			          "\n\tThe rest of other parameters are optional.\n" +
			          "\n\tserver=<server name> " +
			          "(required if middleware is not bolt/MB locally)" +	
			          "\n\tsubject=<subject name>" +
			          "\n\tsubject.<N>=<subject name>" +
			          "\n\t\tNote: N must be 1 or greater, and be consecutive." +
			          "\n\t\tFor example: subject.1=GMSEC.A subject.2=GMSEC.B ..." +
			          "\n\titerations=<iterations>" +
			          "\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
			          "\n\tprog_timeout_s=<timeout period (seconds)>" +
			          "\n\t\tNote: msg_timeout_ms must be defined to use prog_timeout_s" +
			          "\n\tloglevel=<log level>" +
			          "\n\tcfgfile=<config_filename.xml>" +
			          "\n\nFor more information, see API User's Guide\n");
		}


		private void CleanUp()
		{
			if(conn != null)
			{
                for (int i = info.Count - 1; i >= 0; i--)
                {
                    Log.Info("Unsubscribing to " + info[i].GetSubject());
                    var temp = info[i];
                    conn.Unsubscribe(ref temp);
                    info.RemoveAt(i);
                }

                conn.Disconnect();

				Connection.Destroy(ref conn);
			}
		}


		private void Run(string[] args)
		{
			config = new Config(args);

			Initialize(config);
			AddToConfigFromFile(config);

			if (IsOptionInValid(config, args, true))
			{
				PrintUsage();
				return;
			}

			// Output GMSEC API version
			Log.Info(Connection.GetAPIVersion());

			try
			{
				// Create the connection
				conn = Connection.Create(config);

				// Connect
				conn.Connect();

				//Determine the subjects to listen too
				List<string> subjects = DetermineSubjects(config);

				// Get example program configurable options
				int iterations     = Get(config, "ITERATIONS", 0);
				int msg_timeout_ms = Get(config, "MSG_TIMEOUT_MS", (int) GMSECWaitDefs.WAIT_FOREVER);
				int prog_timeout_s = Get(config, "PROG_TIMEOUT_S", (int) GMSECWaitDefs.WAIT_FOREVER);

				if (iterations > 0)
				{
					Log.Info("Waiting for up to " + iterations + " messages");
				}

				Log.Info("Middleware version = " + conn.GetLibraryVersion());

				//Subscribe
				for (int i = 0; i < subjects.Count; i++)
				{
					Log.Info("Subscribing to "+subjects[i]);
					info.Add(conn.Subscribe(subjects[i]));
				}

				//Wait & Print Out Messages
				bool done = false;
				int count = 0;
				Stopwatch stopWatch = new Stopwatch();

				while (!done)
				{
					if (prog_timeout_s != (int) GMSECWaitDefs.WAIT_FOREVER)
					{
						if (!stopWatch.IsRunning)
						{
							stopWatch.Start();
						}
						else if (stopWatch.ElapsedMilliseconds >= (long)(prog_timeout_s * 1000))
						{
							Log.Info("Program Timeout Reached!");
							stopWatch.Stop();
							break;
						}
					}

					Message msg = conn.Receive(msg_timeout_ms);

					if (msg == null)
					{
						Log.Warning("timeout occurred");
						continue;
					}

					++count;

					if (iterations > 0 && count >= iterations)
					{
						done = true;
					}

					// Display XML representation of the received message
					Log.Info("Received:\n" + msg.ToXML());

					string subject = msg.GetSubject();

					if (subject.Equals("GMSEC.TERMINATE"))
					{
						done = true;
					}

					// Destroy message
					conn.Release(ref msg);
				}
			}
			catch (GMSEC_Exception e)
			{
				Log.Error(e.ToString());
			}
			finally
			{
				CleanUp();
			}
		}


		static void Main(string[] args)
		{
			new gmsub().Run(args);
		}
	}
}

File: examples/cpp/gmsub/gmsub.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmsub.cpp
 * 
 * A C++ example demonstration of GMSEC publish/subscribe functionality.
 * The associated example gmsub.cpp will publish a message and this program
 * will subscribe and listen for it.
 * gmsub should be run before gmpub.
 * 
 */

#include "../example.h"

#include <gmsec4/util/TimeUtil.h>

#include <string>
#include <vector>

using namespace gmsec::api;
using namespace gmsec::api::util;


class gmsub
{
public:
	gmsub(Config& config);
	~gmsub();
	bool run();

private:
	typedef std::vector<std::string> Subjects;

	Config&     config;
	Connection* connection;
	Subjects    subjects;
	SubscriptionInfo** info;
};


gmsub::gmsub(Config &c)
	: config(c),
	  connection(0),
	  subjects(0)
{
	// Initialize config
	example::initialize(c);
}


gmsub::~gmsub()
{
	if (connection)
	{
		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Unsubscribing from " << subjects[i].c_str();
			connection->unsubscribe(info[i]);
		}
		delete[] info;
		connection->disconnect();
		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmsub::run()
{
	bool success = true;

	//o output GMSEC API version
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o Determine the subjects to listen to
		example::determineSubjects(config, subjects);

		//o Determine runtime settings
		int iterations     = example::get(config, "ITERATIONS", 0);
		int msg_timeout_ms = example::get(config, "MSG_TIMEOUT_MS", GMSEC_WAIT_FOREVER);
		int prog_timeout_s = example::get(config, "PROG_TIMEOUT_S", GMSEC_WAIT_FOREVER);

		if (iterations > 0)
		{
			GMSEC_INFO << "Waiting for up to " << iterations << " messages";
		}

		//o Output the middleware information
		GMSEC_INFO << "Middleware version = " << connection->getLibraryVersion();

		//o Subscribe
		info = new SubscriptionInfo*[subjects.size()];
		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Subscribing to " << subjects[i].c_str();
			info[i] = connection->subscribe(subjects[i].c_str());
		}

		//o Wait for, and print out messages
		int    count = 0;
		bool   done = false;
		double prevTime;
		double nextTime;
		double elapsedTime = 0;

		prevTime = TimeUtil::getCurrentTime_s();

		while (!done)
		{
			if (prog_timeout_s != GMSEC_WAIT_FOREVER && elapsedTime >= prog_timeout_s)
			{
				GMSEC_INFO << "Program Timeout Reached!";
				done = true;
				continue;
			}

			Message* message = connection->receive(msg_timeout_ms);

			if (prog_timeout_s != GMSEC_WAIT_FOREVER)
			{
				nextTime = TimeUtil::getCurrentTime_s();
				elapsedTime += (nextTime - prevTime);
				prevTime = nextTime;
			}

			if (!message)
			{
				GMSEC_INFO << "timeout occurred";
			}
			else
			{
				GMSEC_INFO << "Received:\n" << message->toXML();

				++count;

				done = (iterations > 0 && count >= iterations);
				done = done || (std::string(message->getSubject()) == "GMSEC.TERMINATE");

				connection->release(message);
			}
		}
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc, "gmsub"))
	{
		example::printUsage("gmsub");
		return -1;
	}

	gmsub(config).run();
}
